File: solutions/cpp/affine-cipher/2/affine_cipher.cpp
#include "affine_cipher.h"

#include <algorithm>
#include <cctype>
#include <stdexcept>
#include <map>

using namespace std;

namespace affine_cipher {

// pairs of valid a's and their modular inverses (mod 26)
    map<int,int> keys_and_inverses  = {
    {1,1}, {3, 9}, {5,21}, {7,15}, {9,3}, {11,19}, {15,7}, {17,23}, {19,11}, {21,5}, {23,17}, {25,25}};
   

    string cipher(string msg, int a, int b, bool decoding){ 
        //valid keys?
          auto search = keys_and_inverses.find(a);
          if (search == keys_and_inverses.end()){
               throw std::invalid_argument("a must be coprime to 26");
          }
        // remove spaces and punctuation
          string out = msg;
          auto new_end = remove_if(out.begin(), out.end(), [](unsigned char ch){return isspace(ch) || ispunct(ch);});
         out.erase(new_end, out.end());
        // encode
        if(!decoding){
           transform(out.begin(), out.end(), out.begin(), [a,b](unsigned char ch){
              return isalpha(ch) ? 
                  ((tolower(ch)-'a')*a + b) % 26 + 'a' : ch; });
        } else { // or decode
            int ia = search->second;
            transform(out.begin(), out.end(), out.begin(), [ia,b](unsigned char ch){
                if (isdigit(ch)) {
                    return ch; }
                else {
                   int de = 
                        (ia*((tolower(ch)-'a') - b))% 26;
                    while (de < 0)
                        de += 26;
                    return (unsigned char)('a' + de);
                }});
        }
        return out;
    }

    string encode(string plaintext, int a, int b){
        string str = cipher(plaintext, a, b, false);
        // group
        for(auto idx = str.begin() + 5; idx < str.end();  idx += 6){
            str.insert(idx, ' ');
        }
        return str;
    }

    
    
    string decode(string ciphertext, int a, int b){      
        return cipher(ciphertext, a, b, true);
    }

}  // namespace affine_cipher

File: solutions/python/affine-cipher/2/affine_cipher.py
def transform(ch,a,b):
    d = ((ord(ch.lower()) - ord('a'))*a + b)%26
    return chr(d + ord('a'))

def inv_transform(ch,a,b):
    # modular inverses for 2*idx+1, i.e. 1,3,5,...  (not for 13)
    inverses = [1, 9, 21, 15, 3, 19, 0, 7, 23, 11, 5, 17, 25]
    ia = inverses[((a-1)//2)]
    d = ia*((ord(ch.lower()) - ord('a'))-b)%26
    if d<0:
        d += 26
    return chr(d + ord('a'))

def chunk_k(str,k):
    if len(str)<=k:
        return str
    else:
        return str[0:k] + ' ' + chunk_k(str[k:],k)

def apply_code(txt, a, b, mode):
     if a%2==0 or a%13==0:
        raise ValueError("a and m must be coprime.")
     out = ""
     for c in txt:
        if c.isdigit():
            out += c
        elif c.isalpha():
            if mode=="en":
                out += transform(c, a, b)  
            else:
                out += inv_transform(c, a, b)
     if mode=="en":
        return chunk_k(out,5)
     else:
        return out    
    
def encode(plain_text, a, b):
    return apply_code(plain_text, a, b, "en")
    
def decode(ciphered_text, a, b):
    return apply_code(ciphered_text, a, b, "de")